keps/sig-auth/3257-trust-anchor-sets/README.md

```diff
@@ -0,0 +1,998 @@
+volume source writes the certificates from a ClusterTrustBundle into the
+container filesystem, with the contents of the projected files updating as the
+corresponding trust anchor sets are updated.
+
+### User Stories
+
+#### Trust Anchor Distribution for Private CAs
+
+In my cluster, I operate a certificates/v1 signer implementation called
+`example.com/server-tls`.  This is backed by a private CA hierarchy.  By some
+unspecified mechanism, I issue certificates from this CA hierarchy to a server
+workload (a pod named `server/server`).  My client workload (a pod named
+`client/client`) needs access to the trust anchors for the CA in order to
+connect to the server.  The client should be able to continually establish
+connections to the server without downtime, even as I rotate the trust anchors
+that back the CA.
+
+To do this, I update the controller that implements `example.com/server-tls` to
+create a ClusterTrustBundle object like this one:
+```yaml
+apiVersion: v1alpha1
+kind: ClusterTrustBundle
+metadata:
+  name: example.com--server-tls-live
+spec:
+  signerName: example.com/server-tls
+  pemTrustAnchors: "<... PEM DATA ...>"
+```
+
+As the set of trust anchors for my CA changes due to rotations and revocations,
+my controller updates the `spec` of the ClusterTrustBundle object.
+
+I then update my client pod to add a `trustAnchors` volume:
+```diff
+apiVersion: v1
+kind: Pod
+metadata:
+  namespace: client
+  name: client
+spec:
+  containers:
+    name: main
+    image: my-image
+    volumeMounts:
++    - mountPath: /var/run/example-com-server-tls-trust-anchors
++      name: example-com-server-tls-trust-anchors
++      readOnly: true
+  volumes:
++  - name: example-com-server-tls-trust-anchors
++    projected:
++      sources:
++      - pemTrustAnchors:
++          clusterTrustBundleName: example.com--server-tls-live
++          path: ca_certificates.pem
+```
+
+When my client pod is scheduled onto a node, Kubelet writes the chosen trust
+anchor set into the container filesystem of my main container.  My main
+container doesn't start until `ca_certificates.pem` is present and non-empty.
+
+I ensure that my client application is capable of reading a set of trust anchors
+from a file containing PEM-wrapped DER-formatted X.509 certificates at the
+location I've specified
+(`/var/run/example-com-server-tls-trust-anchors/ca_certificates.pem`).
+
+I ensure that my client application notices changes to `ca_certificates.pem`
+(using inotify or periodic polling), and properly updates its TLS stack when
+that happens.
+
+### Risks and Mitigations
+
+Scalability: In the limit, ClusterTrustBundle objects will be used by
+every pod in the cluster, which will require watches from all Kubelets.  When
+they are updated, workloads will need to receive the updates fairly quickly
+(within 5 minutes across the whole cluster), to accommodate emergency rotation
+of trust anchors for a private CA.
```

Comment:
I missed the original comment in:
https://github.com/kubernetes/enhancements/pull/3258/#discussion_r986260905
[Now catching up with some proposals that were merged in 1.26 timeframe]


IIUC, the proposal as of now claims, that each Kubelet will have *a single* watch, watching for all ClusterTrustBundle objects. Do I understand correctly?
If so, assuming that the changes to CTBs are infrequent, that sounds fine.

If I misunderstood it and you want to have a per-pod watch(es) for all CTBs it requires, that will for sure cause as a lot of problems, and I we have to revisit this decision.


[In general, high number of watches was hugely problematic in the past, and we don't want to make that even worse with this feature.]

@ahmedtd @mikedanese 
@kubernetes/sig-scalability - FYI

Reply:
yes, intent is single watch per kubelet

@ahmedtd, mind opening an update to clarify that?

Reply:
Thanks Jordan - that sgtm.
Let's just add it to the KEP.

Reply:
Clarified in an update.